pythainlp/transliterate/thai2rom.py

```diff
@@ -50,102 +69,276 @@ def _prepare_sequence_in(self, input_text):
-        self.hidden_dim = hidden_dim
-        self.char_emb = nn.Embedding(vocab_size, emb_dim)
-        self.lstm = nn.LSTM(emb_dim, self.hidden_dim // 2, bidirectional=True)
-
-    def forward(self, input_seq):
-        self.hidden = self.init_hidden()
-        embedded = self.char_emb(input_seq)
-        output, self.hidden = self.lstm(
-            embedded.view(len(embedded), 1, -1), self.hidden)
-        return output, self.hidden
-
-    def init_hidden(self):
-        # The axes semantics are (num_layers, minibatch_size, hidden_dim)
-        return (
-            torch.zeros(
-                2,
-                1,
-                self.hidden_dim // 2,
-                requires_grad=True).to(device),
-            torch.zeros(
-                2,
-                1,
-                self.hidden_dim // 2,
-                requires_grad=True).to(device))
-
-
-class OneStepDecoder(nn.Module):
-    def __init__(self, vocab_size, hidden_dim, emb_dim):
-        super(OneStepDecoder, self).__init__()
-        self.hidden_dim = hidden_dim
-        self.char_emb = nn.Embedding(vocab_size, emb_dim)
-        self.lstm = nn.LSTM(emb_dim, hidden_dim)
-        self.out = nn.Linear(hidden_dim, vocab_size)
-        self.softmax = nn.LogSoftmax(dim=1)
-
-    def forward(self, input_step, hidden, encoder_outputs):
-
-        embedded = self.char_emb(input_step).view(1, 1, -1)
-        output = embedded
-        output = F.relu(output)
-        output, hidden = self.lstm(output, hidden)
-        output = F.log_softmax(self.out(output[0]), dim=1)
-        return output, hidden, []  # this empty list should be replaced with decoder attn score
-
-    def init_hidden(self):
-        # The axes semantics are (num_layers, minibatch_size, hidden_dim)
-        return (
-            torch.zeros(
-                1,
-                1,
-                self.hidden_dim,
-                requires_grad=True).to(device),
-            torch.zeros(
-                1,
-                1,
-                self.hidden_dim,
-                requires_grad=True).to(device))
+        self.hidden_size = hidden_size
+        self.character_embedding = nn.Embedding(vocabulary_size,
+                                                embedding_size)
+        self.rnn = nn.LSTM(
+            input_size=embedding_size,
+            hidden_size=hidden_size // 2,
+            bidirectional=True,
+            batch_first=True,
+        )
+
+        self.dropout = nn.Dropout(dropout)
+
+    def forward(self, sequences, sequences_lengths):
+
+        # sequences: (batch_size, sequence_length=MAX_LENGTH)
+        # sequences_lengths: (batch_size)
+
+        batch_size = sequences.size(0)
+        self.hidden = self.init_hidden(batch_size)
+
+        sequences_lengths = np.sort(sequences_lengths)[::-1]
+        index_sorted = np.argsort(
+            -sequences_lengths
+        )  # use negation in sort in descending order
+        index_unsort = np.argsort(index_sorted)  # to unsorted sequence
```

Comment:
Hi,

May I ask you what you're trying to do in these two lines above?

Reply:
The two lines above are to initialize the hidden state and cell state of the first time step `(h_0, c_0)` to zero. 

However, I think that those two lines and the method `init_hiden` could be removed because if (h_0, c_0) is not provided, both h_0 and c_0 default to zero according to the PyTorch documentation.